### PR TITLE
Fourier DS Init

### PR DESCRIPTION
**Pull request by Sair-Shaikh**
This pull request includes significant updates to the `src/miexp/bfuncs.py` and `src/miexp/models/interptransformer.py` files, focusing on dataset generation and transformer model enhancements. The most important changes include the introduction of a new dataset class, modifications to data type handling, and the addition of new neural network modules.

### Dataset Enhancements:

* [`src/miexp/bfuncs.py`](diffhunk://#diff-e602b46d1f96445fb159645ab03db83be81519f1f9f42c91c2995784c197e47cR93-R139): Added a new class `MultiComponentSpectrumDataset` to generate a dataset for the two-component spectrum boolean function. This includes methods for initialization, length, item retrieval, and label generation.
* [`src/miexp/bfuncs.py`](diffhunk://#diff-e602b46d1f96445fb159645ab03db83be81519f1f9f42c91c2995784c197e47cL81-R82): Changed the data type for the `data` attribute in the existing dataset class from `torch.float` to `torch.int` to improve efficiency.

### Transformer Model Updates:

* [`src/miexp/models/interptransformer.py`](diffhunk://#diff-41e7887abf14a10095f58c462987c4c1ed092c5928d7d4356de4cb6d8ba3afa5R1-R46): Introduced the `MultiHeadAttention` class, which applies multiple attention heads followed by a multi-layer perceptron (MLP). This includes initialization and forward pass methods.
* [`src/miexp/models/interptransformer.py`](diffhunk://#diff-41e7887abf14a10095f58c462987c4c1ed092c5928d7d4356de4cb6d8ba3afa5L39-R161): Added the `PositionalEncoding` class to add positional encoding to the input, enhancing the model's ability to handle sequential data.
* [`src/miexp/models/interptransformer.py`](diffhunk://#diff-41e7887abf14a10095f58c462987c4c1ed092c5928d7d4356de4cb6d8ba3afa5L39-R161): Created the `MultiHeadTransformerPosEmb` class, which combines embedding, positional encoding, multiple attention heads, and unembedding to form a comprehensive transformer model.

### CODE REVIEW

File: src/miexp/bfuncs.py [hunk 1]
```diff
@@ -89,3 +90,52 @@ def __len__(self) -> int:  # noqa: D105
...
+        self.labels = self.generate_labels(self.data, coeffs, comps).type(torch.int)
+
+    def __len__(self) -> int:  # noqa: D105
+        return self.num_samples
+
+    def __getitem__(self, idx: int) -> tuple[torch.Tensor, torch.Tensor]:  # noqa: D105
+        return (self.data[idx], self.labels[idx])
+
+    def generate_labels(self, data: Tensor, coeffs: Tensor, comps: Tensor) -> Tensor:
+        """Generate the labels for the data.
```
Comment (henryscheible) [inline src/miexp/bfuncs.py:129]: Write a better description of data

File: src/miexp/bfuncs.py [hunk 2]
```diff
@@ -89,3 +90,52 @@ def __len__(self) -> int:  # noqa: D105
...
+
+        Args:
+            data (Tensor): Data to generate the labels for
+            coeffs (Tensor): Coeffients for the components
```
Comment (henryscheible) [inline src/miexp/bfuncs.py:133]: Add shapes for all of these tensors

File: experiments/fourier_training/Makefile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+.PHONY: train
+
+train:
+	uv run python train_transformer.py --device cpu --lr 0.1 --wd 1e-5 --dataset_size 1000 \
+	--func_width 20 --head_dim 32 --num_heads 16 --hidden_dim 32 \
+	--visualization_save_path ./vis.pdf --csv_save_path ./results.csv --model_save_path ./model.pth \
+	--dataset_save_path ./dataset.pt --num_epochs 300 --train_frac 0.8
```
Comment (henryscheible) [inline experiments/fourier_training/Makefile:7]: Add a dry run version of this (just scale num epochs to 1 and decrease other hyperparameters) that we can run in CI, and add it to the check-experiments workflow

File: src/miexp/models/interptransformer.py [hunk 4]
```diff
@@ -1,10 +1,49 @@
...
+
+        Attributes:
+            attention_heads (nn.ModuleList): A list of attention heads.
+            mlp (nn.Sequential): A multi-layer perceptron with two linear layers and a ReLU activation in between.
+        """
+        super().__init__()
+        self.attention_heads = nn.ModuleList(
+            [AttentionHead(head_dim, hidden_dim) for _ in range(n_heads)]
+        )
+        self.w_o = nn.Linear(n_heads * head_dim, hidden_dim, bias=False)
```
Comment (henryscheible) [inline src/miexp/models/interptransformer.py:29]: Remove w_o's

**Review by henryscheible — CHANGES_REQUESTED** [review]
Left some comments

**Review by henryscheible — APPROVED** [review]
Looks good!
